net: lwm2m: support client-initiated De-register

junqingzou opened the pull request:
Add new RD Client API of lwm2m_rd_client_stop() for this
Fix issues of de-register and event report in RD Client

jukkar commented on subsys/net/lib/lwm2m/lwm2m_rd_client.c:
```diff
@@ -349,12 +349,12 @@ static int do_deregister_reply_cb(const struct coap_packet *response,
 
 	code = coap_header_get_code(response);
-	LOG_DBG("Deregister callback (code:%u.%u)",
+	LOG_INF("Deregister callback (code:%u.%u)",
```
IMHO these should be LOG_DBG() as typically info level prints just too much data.

jukkar commented on subsys/net/lib/lwm2m/lwm2m_rd_client.c:
```diff
@@ -349,12 +349,12 @@ static int do_deregister_reply_cb(const struct coap_packet *response,
 
 	if (code == COAP_RESPONSE_CODE_DELETED) {
-		LOG_DBG("Deregistration success");
+		LOG_INF("Deregistration success");
```
same here

junqingzou replied:
Thanks for the comment. I'm thinking that it'll be good to have de-register logging at the same level as register (as in "do_registration_reply_cb()"). App developer does not want debug prints but would be nice to know the de-register result.

jukkar replied:
Sure, if LOG_INF() works here and does not print too much data, then it is ok. It is just that usually the info level is printed too much which is quite annoying.

mike-scott reviewed (changes requested):
I think if we move the `lwm2m_engine_context_close` call and fix the minor space NIT we're in a good place for this PR.

junqingzou commented on subsys/net/lib/lwm2m/lwm2m_rd_client.c:
```diff
@@ -844,6 +848,7 @@ static void lwm2m_rd_client_service(struct k_work *work)
 			break;
 
 		case ENGINE_DEREGISTERED:
+			lwm2m_engine_context_close(client.ctx);
```
@mike-scott thanks for the comments, I updated the code accordingly.
>move it to do_deregister_reply_cb() before the line below
Sorry I moved it to "after" the line, did not read the comment carefully enough

mike-scott reviewed (approved):
Thanks for the update @junqingzou: LGTM

pfalcon reviewed (commented):
Is this correct: `RD (De-Registration)`? (Mind the order of letters in abbreviation.)

junqingzou committed:
net: lwm2m: support client-initiated De-register

Add new RD Client API of lwm2m_rd_client_stop() for this
Fix issues of de-register and event reporting in RD Client

Signed-off-by: Jun Qing Zou <jun.qing.zou@nordicsemi.no>

junqingzou commented:
@pfalcon Thanks, updated to use the "De-register" term as in the spec.

rlubos reviewed (approved):
LGTM
